File: src/Plotly.NET/Traces/ObjectAbstractions/Dimensions.fs
namespace Plotly.NET.TraceObjects

open Plotly.NET
open Plotly.NET.LayoutObjects
open DynamicObj
open System
open System.Runtime.InteropServices

/// Dimensions type inherits from dynamic object
type Dimension() =
    inherit DynamicObj()

    /// Initializes dimensions object to be used with parcats and parcoords plots
    static member initParallel
        (
            ?ConstraintRange: StyleParam.Range,
            ?Label: #IConvertible,
            ?MultiSelect: bool,
            ?Name: string,
            ?Range: StyleParam.Range,
            ?TemplateItemName: string,
            ?TickFormat: StyleParam.TickMode,
            ?TickText: seq<#IConvertible>,
            ?Tickvals: seq<#IConvertible>,
            ?Values: seq<#IConvertible>,
            ?Visible: bool
        ) =
        Dimension()
        |> Dimension.style (
            ?ConstraintRange = ConstraintRange,
            ?Label = Label,
            ?MultiSelect = MultiSelect,
            ?Name = Name,
            ?Range = Range,
            ?TemplateItemName = TemplateItemName,
            ?TickFormat = TickFormat,
            ?TickText = TickText,
            ?Tickvals = Tickvals,
            ?Values = Values,
            ?Visible = Visible

        )


    /// Initializes dimensions object to be used with SPLOM plots
    static member initSplom
        (
            ?AxisMatches: bool,
            ?AxisType: StyleParam.AxisType,
            ?Label: #IConvertible,
            ?Name: string,
            ?TemplateItemName: string,
            ?Values: seq<#IConvertible>,
            ?Visible: bool
        ) =
        Dimension()
        |> Dimension.style (
            ?AxisMatches = AxisMatches,
            ?AxisType = AxisType,
            ?Label = Label,
            ?Name = Name,
            ?TemplateItemName = TemplateItemName,
            ?Values = Values,
            ?Visible = Visible

        )


    // Applies the styles to Dimensions()
    static member style
        (
            ?Label: #IConvertible,
            ?Name: string,
            ?TemplateItemName: string,
            ?Values: seq<#IConvertible>,
            ?Visible: bool,
            ?ConstraintRange: StyleParam.Range,
            ?MultiSelect: bool,
            ?Range: StyleParam.Range,
            ?TickFormat: StyleParam.TickMode,
            ?TickText: seq<#IConvertible>,
            ?Tickvals: seq<#IConvertible>,
            ?AxisMatches: bool,
            ?AxisType: StyleParam.AxisType
        ) =
        fun (dims: Dimension) ->

            let axis =
                LinearAxis.init (?AxisType = AxisType)
                |> DynObj.withOptionalProperty "matches" AxisMatches

            dims
            |> DynObj.withOptionalProperty "label" Label
            |> DynObj.withOptionalProperty "name" Name
            |> DynObj.withOptionalProperty "templateitemname" TemplateItemName
            |> DynObj.withOptionalProperty "values" Values
            |> DynObj.withOptionalProperty "visible" Visible
            |> DynObj.withOptionalPropertyBy "constraintrange" ConstraintRange StyleParam.Range.convert
            |> DynObj.withOptionalProperty "multiselect" MultiSelect
            |> DynObj.withOptionalPropertyBy "range" Range StyleParam.Range.convert
            |> DynObj.withOptionalPropertyBy "tickformat" TickFormat StyleParam.TickMode.convert
            |> DynObj.withOptionalProperty "ticktext" TickText
            |> DynObj.withOptionalProperty "tickvals" Tickvals
            |> DynObj.withProperty "axis" axis

File: src/Plotly.NET.CSharp/ObjectAbstractions/TraceObjects/Dimensions.cs

using Microsoft.FSharp.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection.Emit;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;

namespace Plotly.NET.CSharp;

public static partial class TraceObjects
{
    public static Plotly.NET.TraceObjects.Dimension InitParallelDimension<LabelType, TickType, ValuesType>
        (
            Optional<StyleParam.Range> ConstraintRange = default,
            Optional<LabelType> Label = default,
            Optional<bool> MultiSelect = default,
            Optional<string> Name = default,
            Optional<StyleParam.Range> Range = default,
            Optional<string> TemplateItemName = default,
            Optional<StyleParam.TickMode> TickFormat = default,
            Optional<IEnumerable<TickType>> TickText = default,
            Optional<IEnumerable<TickType>> Tickvals = default,
            Optional<IEnumerable<ValuesType>> Values = default,
            Optional<bool> Visible = default
        ) 
            where LabelType : IConvertible
            where TickType : IConvertible
            where ValuesType : IConvertible
            =>
                Plotly.NET.TraceObjects.Dimension.initParallel<LabelType, TickType, TickType, ValuesType>(
                    ConstraintRange: ConstraintRange.ToOption(),
                    Label: Label.ToOption(),
                    MultiSelect: MultiSelect.ToOption(),
                    Name: Name.ToOption(),
                    Range: Range.ToOption(),
                    TemplateItemName: TemplateItemName.ToOption(),
                    TickFormat: TickFormat.ToOption(),
                    TickText: TickText.ToOption(),
                    Tickvals: Tickvals.ToOption(),
                    Values: Values.ToOption(),
                    Visible: Visible.ToOption()
                );

    public static Plotly.NET.TraceObjects.Dimension InitSplomDimension<LabelType, ValuesType>
    (
        Optional<bool> AxisMatches = default,
        Optional<StyleParam.AxisType> AxisType = default,
        Optional<LabelType> Label = default,
        Optional<string> Name = default,
        Optional<string> TemplateItemName = default,
        Optional<IEnumerable<ValuesType>> Values = default,
        Optional<bool> Visible = default
    )
        where LabelType : IConvertible
        where ValuesType : IConvertible
        =>
            Plotly.NET.TraceObjects.Dimension.initSplom<LabelType, ValuesType>(
                AxisMatches: AxisMatches.ToOption(),
                AxisType: AxisType.ToOption(),
                Label: Label.ToOption(),
                Name: Name.ToOption(),
                TemplateItemName: TemplateItemName.ToOption(),
                Values: Values.ToOption(),
                Visible: Visible.ToOption()
            );
}

